### PR TITLE
Silence failing raycast unit test

### PR DESCRIPTION
**Pull request by appgurueu**
The cause for the test failure is an edge case bug in the raycast implementation (perfectly diagonal raycasts).

This is done by switching to a continuous random distribution which makes it extremely unlikely that the buggy edge case occurs.

Additionally, devtest unit test failures now print their random seed to be easier to reproduce in the future.

## How to test

Hacky patch for in-world visualization:

```diff
diff --git a/games/devtest/mods/unittests/raycast.lua b/games/devtest/mods/unittests/raycast.lua
index 533e172d6..149cc0218 100644
--- a/games/devtest/mods/unittests/raycast.lua
+++ b/games/devtest/mods/unittests/raycast.lua
@@ -92,7 +92,34 @@ local function test_raycast_noskip(_, pos)
 		-- or a dirt node got falsely skipped.
 		local pt = core.raycast(ray_start, ray_end, false, false):next()
 		if pt then
-			assert(core.get_node(pt.above).name == "air")
+			if core.get_node(pt.above).name ~= "air" then
+				core.after(1, function()
+					local function mark(pos, color)
+						core.add_particle{
+							pos = pos,
+							expirationtime = 1000,
+							texture = "[fill:1x1:" .. color,
+							size = 1
+						}
+					end
+					for x = 0, 255 do
+						mark((1 - x/255) * ray_start + x/255 * ray_end, ("#%02x0000"):format(x))
+					end
+
+					mark(pt.intersection_point, "green")
+					mark(pt.intersection_point + pt.intersection_normal, "yellow")
+
+					core.swap_node(pt.under, {name = "basenodes:stone"})
+					core.swap_node(pt.above, {name = "basenodes:sand"})
+					
+					print("cuboid", cuboid_minmax(1.5))
+					print("ray", ray_start, ray_end)
+					print("pt", "above", pt.above, "under", pt.under)
+					print("node", core.get_node(pt.above).name)
+				end)
+				return
+			end
+			-- assert(core.get_node(pt.above).name == "air")
 		end
 	end
```
Stone is under, sand is above as you can see. The buggy case looks like this (black is start, red is end of the ray):

Screenshot from 2025-01-06 00-49-26

If you look inside (yellow is intersection point + normal, green is intersection point):

Screenshot from 2025-01-06 00-49-38

### CODE REVIEW

File: games/devtest/mods/unittests/init.lua [hunk 1]
```diff
@@ -119,10 +119,17 @@ local function wait_for_map(pos, callback)
...
+local function printf(fmt, ...)
+	print(fmt:format(...))
+end
+
 -- This runs in a coroutine so it uses await()
 function unittests.run_all()
 	local counters = { time = 0, total = 0, passed = 0 }
 
+	local seed = core.get_us_time()
+	math.randomseed(seed)
```
Comment (sfan5) [inline games/devtest/mods/unittests/init.lua:131]: I fear that in practice everything is too non-deterministic (e.g. some code calls `math.random` in between tests) for this to be really useful.
Conceptually I think we should avoid using random numbers in tests - at least for input data.

Comment (appgurueu) [reply games/devtest/mods/unittests/init.lua:131]: > I fear that in practice everything is too non-deterministic (e.g. some code calls `math.random` in between tests) for this to be really useful.

Fair point. What if we add a `random = true` field to the test def table and if this is set, seed random for each test right before it runs individually? Alternatively we could also pass a `PcgRandom` instance to such tests.

> I think we should avoid using random numbers in tests - at least for input data.

I don't think so. Random numbers are a very powerful tool to test that some assumptions (such as the one unexpectedly violated here) hold for many different configurations with little code.

Comment (sfan5) [reply games/devtest/mods/unittests/init.lua:131]: > What if we add a random = true field to the test def table and if this is set, seed random for each test right before it runs individually?

that sounds good.

File: games/devtest/mods/unittests/init.lua [hunk 2]
```diff
@@ -160,11 +172,11 @@ function unittests.run_all()
...
-	counters.total - counters.passed, counters.total))
-	print(string.format("    Testing took %dms total.", counters.time))
+	local passed = counters.total == counters.passed
+	printf("Devtest Unit Test Results: %s", passed and "PASSED" or "FAILED")
+	printf("    %d / %d failed tests.",
+		counters.total - counters.passed, counters.total)
+	printf("    Testing took %dms total.", counters.time)
 	print(string.rep("+", 80))
 	unittests.on_finished(counters.total == counters.passed)
 	return counters.total == counters.passed
```
Comment (sfan5) [inline games/devtest/mods/unittests/init.lua:182]:
```suggestion
	unittests.on_finished(passed)
	return passed
```

**Review by sfan5 — APPROVED** [review]

**Review by sfan5 — CHANGES_REQUESTED** [review]

**Review by sfan5 — APPROVED** [review]
